pv3_rust/src/pv.rs

```diff
@@ -0,0 +1,395 @@
+    fn new(chunk_size_: u32, chunk_count_: i32) -> PvNic {
+
+        let xsk_ptr;
+        unsafe {
+            let xsk_layout = Layout::new::<xsk_socket>();
+            xsk_ptr = alloc_zeroed(xsk_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let umem_ptr;
+        unsafe {
+            let umem_layout = Layout::new::<xsk_umem>();
+            umem_ptr = alloc_zeroed(umem_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let fq_ptr;
+        unsafe {
+            let fq_layout = Layout::new::<xsk_ring_prod>();
+            fq_ptr = alloc_zeroed(fq_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let rx_ptr;
+        unsafe {
+            let rx_layout = Layout::new::<xsk_ring_cons>();
+            rx_ptr = alloc_zeroed(rx_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let cq_ptr;
+        unsafe {
+            let cq_layout = Layout::new::<xsk_ring_cons>();
+            cq_ptr = alloc_zeroed(cq_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let tx_ptr;
+        unsafe {
+            let tx_layout = Layout::new::<xsk_ring_prod>();
+            tx_ptr = alloc_zeroed(tx_layout);
+            // TODO: Add if umem_ptr.is_null() to make it safe
+        }
+
+        let a:PvNic = unsafe {
+            PvNic {
+                xsk: xsk_ptr.cast::<xsk_socket>(),
+                umem: umem_ptr.cast::<xsk_umem>(),   // umem is needed to be dealloc after using packetvisor library.
+                buffer: std::ptr::null_mut(),
+                chunk_size: chunk_size_,
+                chunk_pool: Vec::with_capacity(chunk_count_ as usize),
+                chunk_pool_idx: chunk_count_,
+                fq: std::ptr::read(fq_ptr.cast::<xsk_ring_prod>()),
+                rx: std::ptr::read(rx_ptr.cast::<xsk_ring_cons>()),
+                cq: std::ptr::read(cq_ptr.cast::<xsk_ring_cons>()),
+                tx: std::ptr::read(tx_ptr.cast::<xsk_ring_prod>()),
+                }
+            };
+
+        a
+    }
+}
+
+fn _pv_alloc(nic: &mut PvNic) -> u64 {
+    if nic.chunk_pool_idx > 0 {
+        nic.chunk_pool_idx -= 1;
+        nic.chunk_pool[nic.chunk_pool_idx as usize]
+    } else {
+        return u64::MAX;
+    }
+}
+
+fn _pv_free(nic: &mut PvNic, chunk_addr: u64) {
+    // align **chunk_addr with chunk size
+    let remainder = chunk_addr % (nic.chunk_size as u64);
+    let chunk_addr = chunk_addr - remainder;
+    nic.chunk_pool[nic.chunk_pool_idx as usize] = chunk_addr;
+    nic.chunk_pool_idx += 1;
+}
+
+/* TODO: packet_process() 구현하면서 remove를 사용해서 벡터중 해당 요소만 제거하는 것으로 소스짜기 (WIP)*/
+pub fn pv_free(nic: &mut PvNic, packets: &mut Vec<PvPacket>) {
+    let packet_count: usize = packets.len();
+
+    for i in 0..packet_count {
+        _pv_free(nic, packets[packet_count - 1 - i].private as u64);
```

Comment:
Possible code style:

```suggestion
    for i in 1..=packet_count {
        _pv_free(nic, packets[packet_count - i].private as u64);
```

Reply:
Nice suggestion. it became simpler